File: Python3/103.Binary Tree Zigzag Level Order Traversal.py
"""
@Title: 103. Binary Tree Zigzag Level Order Traversal
@Tag: binary tree
@Date: Mar-01 2020
@Author: ceezyyy
@Difficulty: Medium
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue, res = [root], []
        while queue:
            n, level = len(queue), []
            for i in range(n):
                cur = queue.pop(0)
                level.append(cur.val)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
            res.append(level)
        for i, r in enumerate(res):
            if i % 2 != 0:
                res[i] = res[i][::-1]
        return res


"""
Runtime: 28 ms, faster than 82.40% of Python3 online submissions for Binary Tree Zigzag Level Order Traversal.
Memory Usage: 12.9 MB, less than 100.00% of Python3 online submissions for Binary Tree Zigzag Level Order Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(1)
"""

File: others/cpp/103.Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Source: LeetCode  103. Binary Tree Zigzag Level Order Traversal
 * Author: ceezyyy
 * Date: 10-09-2019
 */


 /*
 Medium

 Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).

 For example:

 Given binary tree [3,9,20,null,null,15,7],
	 3
	/ \
   9  20
	 /  \
	15   7
 
 return its zigzag level order traversal as:

 [
   [3],
   [20,9],
   [15,7]
 ]

 */


#include<vector>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	struct TreeNode {
		int val;
		TreeNode *left;
		TreeNode *right;
		TreeNode(int x) : val(x), left(NULL), right(NULL) {}
	};

	vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
		vector<vector<int>> res;
		queue<TreeNode*> q;
		if (!root)
			return res;
		q.push(root);
		// the similar as level order traversal
		while (!q.empty()) {
			vector<int> temp;
			int size = q.size();
			while (size--) {
				TreeNode* node = q.front();
				q.pop();
				temp.push_back(node->val);
				if (node->left)
					q.push(node->left);
				if (node->right)
					q.push(node->right);
			}
			res.push_back(temp);
		}
		for (int j = 1; j < res.size(); j = j + 2) {
			reverse(res.at(j).begin(), res.at(j).end());
		}
		return res;
	}
};


/**
 * Runtime: 4 ms, faster than 86.78% of C++ online submissions for Binary Tree Zigzag Level Order Traversal.
 * Memory Usage: 13.4 MB, less than 97.67% of C++ online submissions for Binary Tree Zigzag Level Order Traversal.
 */
